go/kbfs/simplefs/uploads.go

```diff
@@ -0,0 +1,259 @@
+// Copyright 2020 Keybase Inc. All rights reserved.
+// Use of this source code is governed by a BSD
+// license that can be found in the LICENSE file.
+
+package simplefs
+
+import (
+	"encoding/base64"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path"
+	"path/filepath"
+	"sync"
+
+	"github.com/keybase/client/go/kbfs/kbfscrypto"
+	"github.com/keybase/client/go/kbfs/libkbfs"
+	"github.com/keybase/client/go/protocol/keybase1"
+	"github.com/pkg/errors"
+	"golang.org/x/net/context"
+)
+
+const (
+	// ulCtxOpID is the display name for the unique operation SimpleFS ID tag.
+	ulCtxOpID = "SFSULID"
+)
+
+// ulCtxTagKey is the type used for unique context tags
+type ulCtxTagKey int
+
+const (
+	// ulCtxIDKey is the type of the tag for unique operation IDs.
+	ulCtxIDKey ulCtxTagKey = iota
+)
+
+type upload struct {
+	dirToDelete *string
+	uploadID    string
+	opid        keybase1.OpID
+
+	state keybase1.UploadState
+}
+
+type uploadManager struct {
+	k         *SimpleFS
+	publisher libkbfs.SubscriptionManagerPublisher
+	cacheDir  string
+
+	lock      sync.RWMutex
+	tempPaths map[string]bool
+	uploads   map[string]upload
+}
+
+func newUploadManager(simpleFS *SimpleFS) *uploadManager {
+	return &uploadManager{
+		k:         simpleFS,
+		publisher: simpleFS.config.SubscriptionManagerPublisher(),
+		cacheDir:  simpleFS.config.KbEnv().GetCacheDir(),
+		tempPaths: make(map[string]bool),
+		uploads:   make(map[string]upload),
+	}
+}
+
+func (m *uploadManager) makeContext() (ctx context.Context) {
+	return libkbfs.CtxWithRandomIDReplayable(context.Background(), ulCtxIDKey, ulCtxOpID, m.k.log)
+}
+
+func (m *uploadManager) makeTempDir() (tempDirPath string, err error) {
+	buf := make([]byte, 12)
+	err = kbfscrypto.RandRead(buf)
+	if err != nil {
+		return "", err
+	}
+	cacheDirAbs, err := filepath.Abs(m.cacheDir)
+	if err != nil {
+		return "", err
+	}
+	tempDirPath, err = ioutil.TempDir(cacheDirAbs, "uploads-")
+	if err != nil {
+		return "", err
+	}
+
+	m.lock.Lock()
+	defer m.lock.Unlock()
+	m.tempPaths[tempDirPath] = true
+
+	return tempDirPath, nil
+}
+
+func (m *uploadManager) doWhileLocked(action func()) {
+	m.lock.Lock()
+	defer m.lock.Unlock()
+	action()
+}
+
+func (m *uploadManager) getUpload(uploadID string) (upload, bool) {
+	m.lock.RLock()
+	defer m.lock.RUnlock()
+	u, ok := m.uploads[uploadID]
+	return u, ok
+}
+
+func (m *uploadManager) waitForDownload(uploadID string) {
+	upload, ok := m.getUpload(uploadID)
+	if !ok {
+		return
+	}
+	defer func() {
+		if upload.dirToDelete == nil {
+			return
+		}
+		if err := os.RemoveAll(*upload.dirToDelete); err != nil {
+			m.k.log.CDebugf(m.makeContext(), "remove temp dir error %s", err)
+		}
+
+	}()
+	defer m.publisher.PublishChange(keybase1.SubscriptionTopic_UPLOAD_STATUS)
```

Comment:
Can move to the top of the previous defer if you want, might save a couple cycles.  But if you think this is cleaner, that's fine with me too.